### PR TITLE
[luci-interpreter] Add StaticMemoryManager

### PR DESCRIPTION
**Pull request by BalyshevArtem**
This pr adds static memory manager to luci-interpreter.

issue: #7522

draft: #7573

ONE-DCO-1.0-Signed-off-by: Artem Balyshev a.balyshev@partner.samsung.com

### CODE REVIEW

File: compiler/luci-interpreter/include/luci_interpreter/StaticMemoryManager.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+// Used for allocations in static buffer, using offsets.
+class StaticMemoryManager : public IMemoryManager
+{
+public:
+  StaticMemoryManager() = delete;
+
+  explicit StaticMemoryManager(uint8_t *buffer_ptr) { _buffer_ptr = buffer_ptr; }
+
+  ~StaticMemoryManager() override { delete[] _buffer_ptr; }
```
Comment (binarman) [inline compiler/luci-interpreter/include/luci_interpreter/StaticMemoryManager.h:33]: Wait, Idea was to use externally allocated memory, so StaticMemoryManager should not delete data (just like `BuddyMemoryManager`)

Comment (BalyshevArtem) [reply compiler/luci-interpreter/include/luci_interpreter/StaticMemoryManager.h]: done

File: compiler/luci-interpreter/include/luci_interpreter/StaticMemoryManager.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+#ifndef LUCI_INTERPRETER_STATIC_MEMORY_MANAGER_H
+#define LUCI_INTERPRETER_STATIC_MEMORY_MANAGER_H
+
+#include "luci_interpreter/MemoryManager.h"
+
+namespace luci_interpreter
+{
+
+// Used for allocations in static buffer, using offsets.
```
[the remark this answers is not in the document]
Comment (BalyshevArtem) [reply compiler/luci-interpreter/include/luci_interpreter/StaticMemoryManager.h]: done

**Review by binarman — APPROVED** [review]
LGTM

File: compiler/luci-interpreter/include/luci_interpreter/StaticMemoryManager.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+namespace luci_interpreter
+{
+
+// Used for allocations in static buffer, using offsets defined in luci model.
+class StaticMemoryManager : public IMemoryManager
+{
+public:
+  StaticMemoryManager() = delete;
+
+  explicit StaticMemoryManager(uint8_t *buffer_ptr) { _buffer_ptr = buffer_ptr; }
```
Comment (m-bronnikov) [inline compiler/luci-interpreter/include/luci_interpreter/StaticMemoryManager.h:31]: (optional) 
```suggestion
  explicit StaticMemoryManager(uint8_t *buffer_ptr) : _buffer_ptr(buffer_ptr) { /* Do nothing */ }
```

**Review by m-bronnikov — APPROVED** [review]
LGTM

File: compiler/luci-interpreter/src/StaticMemoryManager.cpp [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+void StaticMemoryManager::allocate_memory(luci_interpreter::Tensor &tensor)
+{
+  if (!tensor.is_allocatable())
+  {
+    return;
+  }
+  if (tensor.is_data_allocated())
+  {
+    release_memory(tensor);
+  }
```
Comment (m-bronnikov) [inline compiler/luci-interpreter/src/StaticMemoryManager.cpp:31]: I think memory release is redundant for this manager =)

File: compiler/luci-interpreter/include/luci_interpreter/core/Tensor.h [hunk 5]
```diff
@@ -158,6 +158,10 @@ class Tensor
 
   bool is_data_allocated() const { return _data_allocated; }
 
+  int32_t get_offset() { return _offset; }
```
Comment (m-bronnikov) [inline compiler/luci-interpreter/include/luci_interpreter/core/Tensor.h:161]: (optional)
```suggestion
  int32_t get_offset() const { return _offset; }
```

File: compiler/luci-interpreter/include/luci_interpreter/StaticMemoryManager.h (hunk 3 above)
Comment (BalyshevArtem) [reply compiler/luci-interpreter/include/luci_interpreter/StaticMemoryManager.h]: done

File: compiler/luci-interpreter/include/luci_interpreter/core/Tensor.h (hunk 5 above)
Comment (BalyshevArtem) [reply compiler/luci-interpreter/include/luci_interpreter/core/Tensor.h]: done

File: compiler/luci-interpreter/src/StaticMemoryManager.cpp (hunk 4 above)
Comment (BalyshevArtem) [reply compiler/luci-interpreter/src/StaticMemoryManager.cpp]: done

**Review by binarman — DISMISSED** [review]

File: compiler/luci-interpreter/include/luci_interpreter/StaticMemoryManager.h [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+{
+
+// Used for allocations in static buffer, using offsets defined in luci model.
+class StaticMemoryManager : public IMemoryManager
+{
+public:
+  StaticMemoryManager() = delete;
+
+  explicit StaticMemoryManager(uint8_t *buffer_ptr) : _buffer_ptr(buffer_ptr)
+  { /* Do nothing */
```
Comment (m-bronnikov) [inline compiler/luci-interpreter/include/luci_interpreter/StaticMemoryManager.h:32]: Looks strange =)

**Review by m-bronnikov — APPROVED** [review]
LGTM

File: compiler/luci-interpreter/include/luci_interpreter/StaticMemoryManager.h (hunk 3 above)
Comment (BalyshevArtem) [reply compiler/luci-interpreter/include/luci_interpreter/StaticMemoryManager.h]: > Need to run `nnas format`, I think this line will be wrapped
@binarman @m-bronnikov 
I did nnas format and in result it divided the figured brackets on rows

**Review by binarman — APPROVED** [review]

**Pull request merged**